sshuttle/client.py

```diff
@@ -389,11 +390,15 @@ def expire_connections(now, mux):
         del udp_by_src[peer]
     debug3('Remaining UDP channels: %d\n' % len(udp_by_src))
 
+def check_connections_allowed(mux):
     # we also want to close all TCP connections from sources that have expired their lease
+    if not _allowed_targets_modified and not _sources_modified:
+       return
```

Comment:
The full check below has two conditions: (1) tcp_connection_is_allowed, (2) s.ok

This fast-fail check only seems to memoize the check for (1). However, will there ever be a condition where !s.ok but the connection is still in tcp_conns? If so, that connection is not getting cleaned up (but I am guessing eventually will when ACL changes -- which it does on a timer).

Reply:
The reason I added this check is because the original sshuttle code didn't do this at all.   This check was added to cut the connections based on the firewall rules or if the workspace agent has died.   If the redis listener hasn't received a update, it doesn't need to be rechecked.   New connections are always checked.     The !s.ok is dealt with in runonce.